File: src/client/acontext-py/src/acontext/integrations/claude_agent.py
"""Claude Agent SDK integration for Acontext.

Provides :class:`ClaudeAgentStorage` (async-only) that accepts messages
produced by the Claude Agent SDK ``receive_response()`` stream and persists
**only** ``UserMessage`` and ``AssistantMessage`` to Acontext in Anthropic
format via ``client.sessions.store_message(...)``.

Other message types (``SystemMessage``, ``ResultMessage``, ``StreamEvent``)
are used only for session-id resolution and are **never** stored.

Usage::

    from acontext import AcontextAsyncClient
    from acontext.integrations.claude_agent import ClaudeAgentStorage

    client = AcontextAsyncClient(api_key="sk_project_token")
    storage = ClaudeAgentStorage(client=client)

    async for message in claude_client.receive_response():
        await storage.save_message(message)
"""

from __future__ import annotations

import logging
import re
from dataclasses import asdict, is_dataclass
from typing import Any, Callable

from ..async_client import AcontextAsyncClient
from ..errors import APIError

__all__ = ["ClaudeAgentStorage"]

logger = logging.getLogger(__name__)

_UUID_RE = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$",
    re.IGNORECASE,
)


def _is_uuid(value: str) -> bool:
    """Return ``True`` if *value* looks like a UUID v4 (or any UUID)."""
    return bool(_UUID_RE.match(value))


# ---------------------------------------------------------------------------
# Helpers – coerce to dict
# ---------------------------------------------------------------------------


def _to_dict(msg: Any) -> dict:
    """Coerce *msg* to a plain dict.

    Supports:
    * ``dict`` – returned as-is.
    * dataclass instances – converted via ``dataclasses.asdict``.
    """
    if isinstance(msg, dict):
        return msg
    if is_dataclass(msg) and not isinstance(msg, type):
        return asdict(msg)
    # Pydantic or similar with model_dump
    if hasattr(msg, "model_dump"):
        return msg.model_dump()  # type: ignore[union-attr]
    raise TypeError(
        f"Cannot coerce message of type {type(msg).__name__} to dict. "
        "Pass a dict or a dataclass instance."
    )


# ---------------------------------------------------------------------------
# Helpers – message type detection
# ---------------------------------------------------------------------------


def _is_system_message(msg: dict) -> bool:
    """``SystemMessage`` has ``subtype`` and ``data`` keys, no ``content``."""
    return "subtype" in msg and "data" in msg and "content" not in msg


def _is_result_message(msg: dict) -> bool:
    """``ResultMessage`` has ``subtype`` and ``session_id``; no ``content``."""
    return "subtype" in msg and "session_id" in msg and "content" not in msg


def _is_stream_event(msg: dict) -> bool:
    """``StreamEvent`` has ``uuid``, ``session_id``, and ``event``."""
    return "uuid" in msg and "session_id" in msg and "event" in msg


def _is_user_message(msg: dict) -> bool:
    """``UserMessage`` has ``content`` but no ``model``."""
    return "content" in msg and "model" not in msg


def _is_assistant_message(msg: dict) -> bool:
    """``AssistantMessage`` has ``content`` **and** ``model``."""
    return "content" in msg and "model" in msg


# ---------------------------------------------------------------------------
# Helpers – session id extraction
# ---------------------------------------------------------------------------


def _validate_session_id(sid: str | None) -> str | None:
    """Return *sid* if it is a valid UUID, otherwise warn and return ``None``."""
    if sid is None:
        return None
    if not isinstance(sid, str):
        return None
    if not _is_uuid(sid):
        logger.warning(
            "Ignoring non-UUID session_id from Claude stream: %r", sid
        )
        return None
    return sid


def get_session_id_from_message(msg: dict) -> str | None:
    """Try to extract a Claude session id from *msg*.

    Priority:
    1. ``SystemMessage`` with ``subtype == "init"`` → ``data.get("session_id")``.
    2. ``ResultMessage`` / ``StreamEvent`` → ``.get("session_id")``.

    Returns ``None`` when the message does not carry a session id or
    when the extracted value is not a valid UUID format.
    """
    # SystemMessage init
    if _is_system_message(msg):
        if msg.get("subtype") == "init":
            data = msg.get("data")
            if isinstance(data, dict):
                return _validate_session_id(data.get("session_id"))
        return None

    # ResultMessage / StreamEvent
    if _is_result_message(msg) or _is_stream_event(msg):
        sid = msg.get("session_id")
        if not isinstance(sid, str):
            return None
        return _validate_session_id(sid)

    return None


# ---------------------------------------------------------------------------
# Helpers – block conversion (Claude SDK → Anthropic blob)
# ---------------------------------------------------------------------------


def _is_thinking_block(block: dict) -> bool:
    return "thinking" in block and "signature" in block


def _is_tool_use_block(block: dict) -> bool:
    return "id" in block and "name" in block and "input" in block


def _is_tool_result_block(block: dict) -> bool:
    return "tool_use_id" in block


def _is_text_block(block: dict) -> bool:
    return "text" in block


def _normalize_tool_result_content(content: Any) -> str | list[dict]:
    """Normalize ``ToolResultBlock.content`` to a shape accepted by the API.

    * ``None`` → ``""``
    * ``str`` → as-is
    * ``list`` → ``[{"type": "text", "text": item["text"]}]``
    """
    if content is None:
        return ""
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        return [{"type": "text", "text": item.get("text", "")} for item in content]
    return str(content)


def _convert_block(block: dict, *, role: str, include_thinking: bool) -> dict | None:
    """Convert a single Claude SDK content block to an Anthropic content block.

    Returns ``None`` when the block should be skipped.
    """
    # Order: ThinkingBlock → ToolUseBlock → ToolResultBlock → TextBlock

    if _is_thinking_block(block):
        if not include_thinking:
            return None
        thinking_text = block.get("thinking", "")
        if not thinking_text:
            return None  # empty thinking text rejected by API
        return {
            "type": "thinking",
            "thinking": thinking_text,
            "signature": block.get("signature", ""),
        }

    if _is_tool_use_block(block):
        if role != "assistant":
            return None  # tool_use only valid in assistant messages
        input_val = block["input"]
        if isinstance(input_val, str):
            import json as _json

            try:
                input_val = _json.loads(input_val)
            except (ValueError, TypeError):
                input_val = {"raw": input_val}
        return {
            "type": "tool_use",
            "id": block["id"],
            "name": block["name"],
            "input": input_val,
        }

    if _is_tool_result_block(block):
        if role != "user":
            return None  # tool_result only valid in user messages
        result: dict[str, Any] = {
            "type": "tool_result",
            "tool_use_id": block["tool_use_id"],
            "content": _normalize_tool_result_content(block.get("content")),
        }
        if block.get("is_error"):
            result["is_error"] = True
        return result

    if _is_text_block(block):
        text = block.get("text", "")
        if not text:
            return None  # empty text rejected by API
        return {"type": "text", "text": text}

    # Unknown block type – skip silently
    return None


def _convert_content_blocks(
    content: str | list, *, role: str, include_thinking: bool
) -> tuple[list[dict], bool]:
    """Convert Claude SDK content to Anthropic content block array.

    Returns ``(blocks, has_thinking)`` where *has_thinking* is ``True``
    when at least one thinking block was included in the output.
    """
    has_thinking = False

    if isinstance(content, str):
        if not content:
            return [], False
        return [{"type": "text", "text": content}], False

    blocks: list[dict] = []
    for block in content:
        if not isinstance(block, dict):
            continue
        converted = _convert_block(block, role=role, include_thinking=include_thinking)
        if converted is not None:
            blocks.append(converted)
            # Track whether a thinking block was included
            if _is_thinking_block(block) and include_thinking:
                has_thinking = True

    return blocks, has_thinking


# ---------------------------------------------------------------------------
# Public conversion helpers
# ---------------------------------------------------------------------------


def claude_user_message_to_anthropic_blob(msg: dict) -> dict | None:
    """Convert a Claude SDK ``UserMessage`` dict to an Anthropic blob.

    Returns ``None`` when the resulting content would be empty (no storable
    blocks).
    """
    content = msg.get("content", "")
    blocks, _ = _convert_content_blocks(content, role="user", include_thinking=False)
    if not blocks:
        return None
    return {"role": "user", "content": blocks}


def claude_assistant_message_to_anthropic_blob(
    msg: dict, *, include_thinking: bool = False
) -> tuple[dict | None, bool]:
    """Convert a Claude SDK ``AssistantMessage`` dict to an Anthropic blob.

    Returns ``(blob_or_none, has_thinking)``.
    *blob_or_none* is ``None`` when the resulting content would be empty.
    *has_thinking* is ``True`` when thinking blocks were included.
    """
    content = msg.get("content", [])
    blocks, has_thinking = _convert_content_blocks(
        content, role="assistant", include_thinking=include_thinking
    )
    if not blocks:
        return None, has_thinking
    return {"role": "assistant", "content": blocks}, has_thinking


# ---------------------------------------------------------------------------
# ClaudeAgentStorage
# ---------------------------------------------------------------------------


class ClaudeAgentStorage:
    """Async-only storage adapter for the Claude Agent SDK.

    Accepts messages from ``claude_client.receive_response()`` and persists
    **only** ``UserMessage`` and ``AssistantMessage`` to Acontext in Anthropic
    format.

    Parameters
    ----------
    client:
        An :class:`AcontextAsyncClient` instance.
    session_id:
        Acontext session UUID. If ``None``, the session id will be discovered
        from the first ``SystemMessage`` with ``subtype == "init"``, or a new
        Acontext session will be created automatically.
    user:
        Optional user identifier string passed to ``sessions.create()``.
        Associates the Acontext session with this user.
    include_thinking:
        Whether to store ``ThinkingBlock`` content as native thinking blocks.
        Default ``False`` (omit thinking).
    on_error:
        Optional callback ``(exception, msg_dict) -> None`` invoked when
        ``store_message`` raises. If not provided, exceptions are logged
        and swallowed.
    """

    def __init__(
        self,
        client: AcontextAsyncClient,
        session_id: str | None = None,
        user: str | None = None,
        include_thinking: bool = False,
        on_error: Callable[[Exception, dict], None] | None = None,
    ) -> None:
        self._client = client
        self._session_id = session_id
        self._user = user
        self._include_thinking = include_thinking
        self._on_error = on_error
        self._session_ensured = False

    # -- properties ----------------------------------------------------------

    @property
    def session_id(self) -> str | None:
        """The current Acontext session id (may be ``None`` until resolved)."""
        return self._session_id

    # -- internal helpers ----------------------------------------------------

    def _try_update_session_id(self, msg: dict) -> None:
        """Update ``_session_id`` from *msg* if not already set."""
        if self._session_id is not None:
            return
        sid = get_session_id_from_message(msg)
        if sid:
            self._session_id = sid
            logger.debug("Resolved session_id=%s from message", sid)

    # -- public API ----------------------------------------------------------

    async def save_message(self, msg: Any) -> None:
        """Persist a single Claude Agent SDK message to Acontext.

        * ``UserMessage`` and ``AssistantMessage`` are stored.
          Errored ``AssistantMessage`` messages are stored when they
          have valid content; the ``error`` value is included in
          message ``meta`` for observability.
        * ``SystemMessage``, ``ResultMessage``, and ``StreamEvent`` are used
          only for session-id resolution and are **not** stored.
        * If ``session_id`` is unknown when a storable message arrives, the
          message is **skipped with a warning** (no exception raised).
        * API errors are caught and either forwarded to ``on_error`` or
          logged, so the caller's ``async for`` loop is never interrupted.
        """
        msg_dict = _to_dict(msg)

        # -- non-storable message types: update session_id only --------------
        if (
            _is_system_message(msg_dict)
            or _is_result_message(msg_dict)
            or _is_stream_event(msg_dict)
        ):
            self._try_update_session_id(msg_dict)
            return

        # -- storable: user or assistant -------------------------------------
        if _is_assistant_message(msg_dict):
            return await self._store_assistant(msg_dict)

        if _is_user_message(msg_dict):
            return await self._store_user(msg_dict)

        # Unknown message shape – ignore
        logger.debug("Ignoring unknown message shape: %s", list(msg_dict.keys()))

    # -- private store methods -----------------------------------------------

    async def _store_user(self, msg: dict) -> None:
        blob = claude_user_message_to_anthropic_blob(msg)
        if blob is None:
            logger.debug(
                "UserMessage produced empty content after conversion – skipping."
            )
            return

        await self._call_store(blob, meta=None)

    async def _store_assistant(self, msg: dict) -> None:
        blob, has_thinking = claude_assistant_message_to_anthropic_blob(
            msg, include_thinking=self._include_thinking
        )
        if blob is None:
            logger.debug(
                "AssistantMessage produced empty content after conversion – skipping."
            )
            return

        meta: dict[str, Any] = {}
        model = msg.get("model")
        if model:
            meta["model"] = model
        if has_thinking:
            meta["has_thinking"] = True
        error = msg.get("error")
        if error:
            meta["error"] = error

        await self._call_store(blob, meta=meta or None)

    async def _ensure_session(self) -> None:
        """Create the Acontext session if it hasn't been created yet.

        If ``_session_id`` is set, uses ``use_uuid`` so the Acontext session
        matches that id and ignores 409 Conflict (already exists).
        If ``_session_id`` is ``None``, lets Acontext generate the id and
        stores it back into ``_session_id``.
        """
        if self._session_ensured:
            return
        try:
            session = await self._client.sessions.create(
                use_uuid=self._session_id if self._session_id else None,
                user=self._user,
            )
            self._session_id = session.id
            logger.debug("Created Acontext session %s", self._session_id)
        except APIError as exc:
            if exc.status_code == 409:
                logger.debug(
                    "Session %s already exists (409) – continuing.",
                    self._session_id,
                )
            else:
                raise
        self._session_ensured = True

    async def _call_store(self, blob: dict, *, meta: dict[str, Any] | None) -> None:
        """Create session if needed, then call ``store_message`` with error resilience."""
        try:
            await self._ensure_session()
            await self._client.sessions.store_message(
                self._session_id,  # type: ignore[arg-type]
                blob=blob,
                format="anthropic",
                meta=meta,
            )
        except Exception as exc:
            if self._on_error is not None:
                self._on_error(exc, blob)
            else:
                logger.warning(
                    "Failed to store message (session=%s): %s",
                    self._session_id,
                    exc,
                    exc_info=True,
                )

File: src/client/acontext-ts/src/integrations/claude-agent.ts
/**
 * Claude Agent SDK integration for Acontext.
 *
 * Provides {@link ClaudeAgentStorage} that accepts messages produced by the
 * Claude Agent SDK `query()` async iterable and persists **only** user and
 * assistant messages to Acontext in Anthropic format via
 * `client.sessions.storeMessage(...)`.
 *
 * Other message types (system, result, stream_event, etc.) are used only for
 * session-id resolution and are **never** stored.
 *
 * Usage:
 * ```typescript
 * import { AcontextClient, ClaudeAgentStorage } from '@acontext/acontext';
 * import { query } from '@anthropic-ai/claude-agent-sdk';
 *
 * const client = new AcontextClient({ apiKey: 'sk-ac-your-api-key' });
 * const storage = new ClaudeAgentStorage({ client });
 *
 * for await (const message of query({ prompt: 'Hello' })) {
 *   await storage.saveMessage(message);
 * }
 * ```
 */

import { APIError } from '../errors';

// ---------------------------------------------------------------------------
// Duck-typed client interface
// ---------------------------------------------------------------------------

/**
 * Minimal duck-typed interface for the Acontext client.
 * Both `AcontextClient` and `MockAcontextClient` satisfy this interface.
 */
export interface AcontextClientLike {
  sessions: {
    create(options?: {
      useUuid?: string | null;
      user?: string | null;
    }): Promise<{ id: string }>;
    storeMessage(
      sessionId: string,
      blob: Record<string, unknown>,
      options?: {
        format?: string;
        meta?: Record<string, unknown> | null;
      }
    ): Promise<unknown>;
  };
}

// ---------------------------------------------------------------------------
// Options
// ---------------------------------------------------------------------------

/**
 * Options for constructing a {@link ClaudeAgentStorage} instance.
 */
export interface ClaudeAgentStorageOptions {
  /** An Acontext client instance (or any object satisfying `AcontextClientLike`). */
  client: AcontextClientLike;
  /** Acontext session UUID. If omitted, discovered from the Claude stream or auto-created. */
  sessionId?: string;
  /** Optional user identifier passed to `sessions.create()`. */
  user?: string;
  /** Whether to store ThinkingBlock content as native thinking blocks. Default: false. */
  includeThinking?: boolean;
  /**
   * Optional error callback invoked when `storeMessage` raises.
   * Receives `(error, blob)` where `blob` is the **converted** Anthropic blob.
   * If not provided, errors are logged via `console.warn`.
   */
  onError?: (error: Error, blob: Record<string, unknown>) => void;
}

// ---------------------------------------------------------------------------
// Helpers – message type detection (TS Claude Agent SDK uses `type` field)
// ---------------------------------------------------------------------------

/**
 * Check if a message is a storable user message.
 * Skips replay messages (`isReplay === true`) which are TS-only.
 */
function isUserMessage(msg: Record<string, unknown>): boolean {
  return msg.type === 'user' && !msg.isReplay;
}

/**
 * Check if a message is a storable assistant message.
 */
function isAssistantMessage(msg: Record<string, unknown>): boolean {
  return msg.type === 'assistant';
}

/**
 * Check if a message is a replay user message (TS-only).
 */
function isReplayMessage(msg: Record<string, unknown>): boolean {
  return msg.type === 'user' && msg.isReplay === true;
}

// ---------------------------------------------------------------------------
// Helpers – session id extraction
// ---------------------------------------------------------------------------

const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

/**
 * Return `sid` if it is a valid UUID, otherwise warn and return `null`.
 */
function validateSessionId(sid: string): string | null {
  if (!UUID_RE.test(sid)) {
    console.warn(
      `Ignoring non-UUID session_id from Claude stream: "${sid}"`
    );
    return null;
  }
  return sid;
}

/**
 * Try to extract a Claude session id from a non-storable message.
 *
 * In the TS Claude Agent SDK, `session_id` is a flat field on system,
 * result, stream_event, and other non-storable message types.
 *
 * Returns `null` when the message does not carry a session id or
 * when the extracted value is not a valid UUID format.
 */
export function getSessionIdFromMessage(
  msg: Record<string, unknown>
): string | null {
  // Only extract from non-storable messages
  if (msg.type === 'user' || msg.type === 'assistant') {
    return null;
  }
  const sid = msg.session_id;
  if (typeof sid !== 'string') {
    return null;
  }
  return validateSessionId(sid);
}

// ---------------------------------------------------------------------------
// Helpers – block conversion (Claude SDK → Anthropic blob)
// ---------------------------------------------------------------------------

/**
 * Normalize `ToolResultBlock.content` to a shape accepted by the API.
 *
 * - `null`/`undefined` → `""`
 * - `string` → as-is
 * - `array` → `[{ type: "text", text: item.text ?? "" }]`
 * - other → `String(content)`
 */
function normalizeToolResultContent(
  content: unknown
): string | Array<Record<string, unknown>> {
  if (content === null || content === undefined) {
    return '';
  }
  if (typeof content === 'string') {
    return content;
  }
  if (Array.isArray(content)) {
    return content.map((item: Record<string, unknown>) => ({
      type: 'text',
      text: (item as Record<string, unknown>)?.text ?? '',
    }));
  }
  return String(content);
}

/**
 * Convert a single Claude SDK content block to an Anthropic content block.
 *
 * Returns `null` when the block should be skipped.
 */
function convertBlock(
  block: Record<string, unknown>,
  role: string,
  includeThinking: boolean
): Record<string, unknown> | null {
  const blockType = block.type;

  switch (blockType) {
    case 'thinking': {
      if (!includeThinking) {
        return null;
      }
      const thinkingText = block.thinking;
      if (!thinkingText) {
        return null; // empty thinking text
      }
      return {
        type: 'thinking',
        thinking: thinkingText,
        signature: block.signature ?? '',
      };
    }

    case 'tool_use': {
      if (role !== 'assistant') {
        return null; // tool_use only valid in assistant messages
      }
      let inputVal = block.input;
      if (typeof inputVal === 'string') {
        try {
          inputVal = JSON.parse(inputVal);
        } catch {
          inputVal = { raw: inputVal };
        }
      }
      return {
        type: 'tool_use',
        id: block.id,
        name: block.name,
        input: inputVal,
      };
    }

    case 'tool_result': {
      if (role !== 'user') {
        return null; // tool_result only valid in user messages
      }
      const result: Record<string, unknown> = {
        type: 'tool_result',
        tool_use_id: block.tool_use_id,
        content: normalizeToolResultContent(block.content),
      };
      if (block.is_error) {
        result.is_error = true;
      }
      return result;
    }

    case 'text': {
      const text = block.text;
      if (!text) {
        return null; // empty text
      }
      return { type: 'text', text };
    }

    default:
      // Unknown block type – skip silently
      return null;
  }
}

/**
 * Convert Claude SDK content to Anthropic content block array.
 *
 * Returns `[blocks, hasThinking]` where `hasThinking` is `true` when at
 * least one thinking block was successfully included in the output.
 */
function convertContentBlocks(
  content: unknown,
  role: string,
  includeThinking: boolean
): [Array<Record<string, unknown>>, boolean] {
  let hasThinking = false;

  if (typeof content === 'string') {
    if (!content) {
      return [[], false];
    }
    return [[{ type: 'text', text: content }], false];
  }

  if (!Array.isArray(content)) {
    return [[], false];
  }

  const blocks: Array<Record<string, unknown>> = [];
  for (const block of content) {
    if (typeof block !== 'object' || block === null) {
      continue; // skip non-object items (matching Python's `if not isinstance(block, dict)`)
    }
    const converted = convertBlock(
      block as Record<string, unknown>,
      role,
      includeThinking
    );
    if (converted !== null) {
      blocks.push(converted);
      // Track whether a thinking block was successfully included
      if (
        (block as Record<string, unknown>).type === 'thinking' &&
        includeThinking
      ) {
        hasThinking = true;
      }
    }
  }

  return [blocks, hasThinking];
}

// ---------------------------------------------------------------------------
// Public conversion helpers
// ---------------------------------------------------------------------------

/**
 * Convert a Claude Agent SDK user message to an Anthropic blob.
 *
 * TS reads `msg.message?.content ?? ""` (nested under API message object).
 *
 * Returns `null` when the resulting content would be empty (no storable blocks).
 */
export function claudeUserMessageToAnthropicBlob(
  msg: Record<string, unknown>
): Record<string, unknown> | null {
  const message = msg.message as Record<string, unknown> | undefined;
  const content = message?.content ?? '';
  const [blocks] = convertContentBlocks(content, 'user', false);
  if (blocks.length === 0) {
    return null;
  }
  return { role: 'user', content: blocks };
}

/**
 * Convert a Claude Agent SDK assistant message to an Anthropic blob.
 *
 * TS reads `msg.message?.content ?? []` (nested under API message object).
 *
 * Returns `{ blob: ... | null, hasThinking: boolean }`.
 */
export function claudeAssistantMessageToAnthropicBlob(
  msg: Record<string, unknown>,
  includeThinking = false
): { blob: Record<string, unknown> | null; hasThinking: boolean } {
  const message = msg.message as Record<string, unknown> | undefined;
  const content = message?.content ?? [];
  const [blocks, hasThinking] = convertContentBlocks(
    content,
    'assistant',
    includeThinking
  );
  if (blocks.length === 0) {
    return { blob: null, hasThinking };
  }
  return {
    blob: { role: 'assistant', content: blocks },
    hasThinking,
  };
}

// ---------------------------------------------------------------------------
// ClaudeAgentStorage
// ---------------------------------------------------------------------------

/**
 * Storage adapter for the Claude Agent SDK (TypeScript).
 *
 * Accepts messages from the `query()` async iterable and persists **only**
 * user and assistant messages to Acontext in Anthropic format.
 */
export class ClaudeAgentStorage {
  private _client: AcontextClientLike;
  private _sessionId: string | null;
  private _user: string | null;
  private _includeThinking: boolean;
  private _onError:
    | ((error: Error, blob: Record<string, unknown>) => void)
    | null;
  private _sessionEnsured = false;

  constructor(options: ClaudeAgentStorageOptions) {
    this._client = options.client;
    this._sessionId = options.sessionId ?? null;
    this._user = options.user ?? null;
    this._includeThinking = options.includeThinking ?? false;
    this._onError = options.onError ?? null;
  }

  // -- properties ----------------------------------------------------------

  /**
   * The current Acontext session id (may be `null` until resolved).
   */
  get sessionId(): string | null {
    return this._sessionId;
  }

  // -- public API ----------------------------------------------------------

  /**
   * Persist a single Claude Agent SDK message to Acontext.
   *
   * - User and assistant messages are stored.
   * - All other message types (system, result, stream_event, etc.) are used
   *   only for session-id resolution and are **not** stored.
   * - Replay messages (`isReplay: true`) are skipped to prevent duplicates.
   * - API errors are caught and either forwarded to `onError` or logged,
   *   so the caller's `for await` loop is never interrupted.
   */
  async saveMessage(msg: Record<string, unknown>): Promise<void> {
    // -- non-storable message types: update session_id only ----------------
    if (msg.type !== 'user' && msg.type !== 'assistant') {
      this._tryUpdateSessionId(msg);
      return;
    }

    // -- replay user message: skip (TS-only) -------------------------------
    if (isReplayMessage(msg)) {
      return;
    }

    // -- storable: assistant or user ---------------------------------------
    if (isAssistantMessage(msg)) {
      return this._storeAssistant(msg);
    }

    if (isUserMessage(msg)) {
      return this._storeUser(msg);
    }

    // Unknown — ignore
  }

  // -- internal helpers ----------------------------------------------------

  private _tryUpdateSessionId(msg: Record<string, unknown>): void {
    if (this._sessionId !== null) {
      return;
    }
    const sid = getSessionIdFromMessage(msg);
    if (sid) {
      this._sessionId = sid;
      console.debug(`Resolved session_id=${sid} from message`);
    }
  }

  // -- private store methods -----------------------------------------------

  private async _storeUser(msg: Record<string, unknown>): Promise<void> {
    const blob = claudeUserMessageToAnthropicBlob(msg);
    if (blob === null) {
      console.debug(
        'UserMessage produced empty content after conversion – skipping.'
      );
      return;
    }
    await this._callStore(blob, null);
  }

  private async _storeAssistant(msg: Record<string, unknown>): Promise<void> {
    const { blob, hasThinking } = claudeAssistantMessageToAnthropicBlob(
      msg,
      this._includeThinking
    );
    if (blob === null) {
      console.debug(
        'AssistantMessage produced empty content after conversion – skipping.'
      );
      return;
    }

    const meta: Record<string, unknown> = {};
    const message = msg.message as Record<string, unknown> | undefined;
    const model = message?.model;
    if (model) {
      meta.model = model;
    }
    if (hasThinking) {
      meta.has_thinking = true;
    }
    const error = msg.error;
    if (error) {
      meta.error = error;
    }

    await this._callStore(
      blob,
      Object.keys(meta).length > 0 ? meta : null
    );
  }

  private async _ensureSession(): Promise<void> {
    if (this._sessionEnsured) {
      return;
    }
    try {
      const session = await this._client.sessions.create({
        useUuid: this._sessionId ? this._sessionId : null,
        user: this._user,
      });
      this._sessionId = session.id;
      console.debug(`Created Acontext session ${this._sessionId}`);
    } catch (err) {
      if (err instanceof APIError && err.statusCode === 409) {
        console.debug(
          `Session ${this._sessionId} already exists (409) – continuing.`
        );
      } else {
        throw err;
      }
    }
    this._sessionEnsured = true;
  }

  private async _callStore(
    blob: Record<string, unknown>,
    meta: Record<string, unknown> | null
  ): Promise<void> {
    try {
      await this._ensureSession();
      await this._client.sessions.storeMessage(this._sessionId!, blob, {
        format: 'anthropic',
        meta,
      });
    } catch (err) {
      if (this._onError) {
        this._onError(err as Error, blob);
      } else {
        console.warn(
          `Failed to store message (session=${this._sessionId}):`,
          err
        );
      }
    }
  }
}
